### DISCUSSION TITLE
How to debug/log deadlock in fuse mount

### DISCUSSION THREAD

**Discussion by kvaster** [seaweedfs/seaweedfs#4949] [General]
I'm using seaweedfs csi driver compiled with seaweedfs mount from commit 40888b561226004765ce158e922df6aca5dcfe70. And I have reproduced deadlock on wirte 4 times already. My question is what can I do to detect / locate / debug / fix the problem.... Thanks in advance.

**Comment by chrislusf** [comment]
The CIS driver is just a simple wrapper. You can run "weed mount" directly.

You can actually remote connect to a "weed mount" instance, if you "make debug_mount" with the `weed/Makefile` file.
( I use IDE IntelliJ)

**Reply by kvaster** [reply]
The problem is that I can reproduce it only in production with a big load.

**Comment by chrislusf** [comment]
@werdnum has a way to do it:
* https://github.com/seaweedfs/seaweedfs/issues/4405#issuecomment-1519017408
* https://github.com/seaweedfs/seaweedfs/issues/4452
* https://github.com/seaweedfs/seaweedfs/issues/4260

But I do not think the `go-deadlock` has a PR.

**Reply by chrislusf** [reply]
There are several old bugs about the deadlock. But with recent changes, the old bugs may not be accurate any more.

**Reply by werdnum** [reply]
I had a PR somewhere.

In general I generated it by installing `github.com/sasha-s/go-deadlock`, replacing `sync.(RW)?Mutex` with `go_deadlock.\1Mutex` on the command line, and then adding the import to all the files I modified. It didn't take too long, just a bit of sed-fu.

**Reply by kvaster** [reply]
Thanks. I will try to locate the deadlock.

**Comment by kvaster** [comment]
Deadlock located and fix is in testing - will make PR very soon.

https://github.com/seaweedfs/seaweedfs/issues/4953
